File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool check(stack<char>& st, char ch){
        if(!st.empty()){
            if(st.top()=='(' && ch==')')
                return true;
            else if(st.top()=='{' && ch=='}')
                return true;
            else if(st.top()=='[' && ch==']')
                return true;
            else return false;
        }
        return false;

    }
    bool isValid(string s) {
        stack<char> st;
        for(char ch: s){
            if(check(st,ch)){
                st.pop();
            }else{
                st.push(ch);
            }
        }

        return st.empty();
    }
};

auto init = []()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 'c';
}();

File: 0020-valid-parentheses/0020-valid-parentheses.cs
using System;
using System.Collections.Generic;

public class Solution {
    public bool IsValid(string s) {
        if (s.Length % 2 != 0) {
            return false;
        }

        Dictionary<char, char> bracketMap = new Dictionary<char, char>() {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        Stack<char> stack = new Stack<char>();

        foreach (char c in s) {
            if (bracketMap.ContainsKey(c)) {
                // Closing bracket
                if (stack.Count == 0 || stack.Pop() != bracketMap[c]) {
                    return false;
                }
            } else {
                // Opening bracket
                stack.Push(c);
            }
        }

        return stack.Count == 0;
    }
}